### ISSUE TITLE
alternative to -io flags and *input* / babashka.input

### ISSUE THREAD

**Issue by borkdude** [babashka/babashka#613]
TL;DR:

- The benefit of not having flags is that it works better for scripts.
- `user/*input*` will be "replaced" with the delays `babashka.input/lines`, `babashka.input/edn`, `babashka.input/edn-seq` which are referred in the `user` namespace for one-liner usage:

```
 cat http://README.md | bb '(while true (run! println (shuffle @lines)))'
```

- how would this work with `--stream`? it won't, but we won't remove the existing I/O flags

Longer version:

We could have a `babashka.io` namespace that has a versatile `parse` function (not `read` since that conflict with clojure.core).

This should be easy to use from one-liners:

```
bb -e "(parse :edn)"
```

Note that in cases like this:

``` clojure
 cat http://README.md | bb -i '(while true (run! println (shuffle *input*))))'
```
(see https://twitter.com/borkdude/status/1327261838522937350)
it is more convenient to have one value bound to the input once without re-reading it. This won't work since it will keep reading from stdin over and over again.

```
 cat http://README.md | bb -i '(while true (run! println (shuffle (io/parse-seq :lines))))'
```

Instead we could have clearer variable names for each case:

```
 cat http://README.md | bb '(while true (run! println (shuffle *lines*)))'
```

But since `*input*` is not really a dynamic var (or shouldn't have to be) and really a delay, we could also use `@lines`, `@edn-vals` and alias `*input*` to `@input` for consistency (undocumenting `*input*`). We could automatically refer these functions in the `user` namespace for cmd line usage, while their origin is in `babashka.io`. They could then also be made available as a JVM lib.

**Comment by borkdude** [comment]
The benefit of not having flags is that it works better for scripts.

**Comment by borkdude** [comment]
babashka.io helper namespace with parse-edn-stream, etc could work

**Comment by borkdude** [comment]
which then deprecates these flags over time

**Comment by borkdude** [comment]
(babashka.io/edn-stream) (assumes input is *in*)
(babashka.io/edn-stream reader) (explicit reader)
same for json, transit

**Comment by borkdude** [comment]
and these functions could be part of the user namespace for one-liners
echo '{:a 1} {:a 2}' | bb '(edn-stream)'

**Comment by borkdude** [comment]
`echo '{:a 1} {:a 2}' | bb '(edn-stream *in*)'` for more explicitness

**Comment by borkdude** [comment]
And this lib could potentially be used from the jvm as well (I guess). So would make it easier to reuse both directions.

Maybe babashka.formats is a better name?

**Comment by borkdude** [comment]
I think the first step would not be writing a new namespace but provide examples/docs how to write the code without i/o flags.

Some people really like the i/o flags (https://lambdaisland.com/blog/2019-12-05-advent-of-parens-5-clojure-in-the-shell, https://twitter.com/jgomo3/status/1316441226795724807) and it's better to not have any breaking changes.

**Comment by borkdude** [comment]
Docs: https://github.com/borkdude/babashka/blob/master/doc/io-flags.md

**Comment by borkdude** [comment]
`babashka.io`
```
;; read
(read :edn) ;; read single edn value from stdin
(read :edn reader) ;; read single edn value from reader
(read-seq :edn) ;; read lazy seq of edn values from stdin
(read-seq :edn reader) ;; reader lazy seq of edn values from stdin

;; variations:
(read :json)
(read-seq :json)

(read :line)
(read-seq :line)

;; write
(write :json v) ;; write JSON value to stoud
(write :json v writer) ;; write JSON value to writer
```
